Change TPC-H to prepared statements

mrks opened the pull request:
This is the last step before we can enable random queries. For now, it prepares the queries but executes them with the same values.

Surprisingly, we have some performance regressions:

```
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
| Benchmark | prev. iter/s  | runs  | new iter/s    | runs  | change | p-value (significant if <0.001) |
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
| TPC-H 1   | 6.13508749008 | 369   | 6.06086301804 | 364   | -1%    |                          0.0000 |
| TPC-H 2   | 130.748535156 | 7845  | 107.744796753 | 6465  | -18%   |     (run time too short) 0.0000 |
| TPC-H 3   | 65.1095962524 | 3907  | 65.5049591064 | 3931  | +1%    |     (run time too short) 0.0000 |
| TPC-H 4   | 23.7911758423 | 1428  | 23.9849758148 | 1440  | +1%    |     (run time too short) 0.0000 |
| TPC-H 5   | 50.4212188721 | 3026  | 49.3288497925 | 2960  | -2%    |     (run time too short) 0.0000 |
| TPC-H 6   | 190.104202271 | 10000 | 197.283874512 | 10000 | +4%    |     (run time too short) 0.0000 |
| TPC-H 7   | 20.310710907  | 1219  | 20.7915840149 | 1248  | +2%    |     (run time too short) 0.0000 |
| TPC-H 8   | 85.8996963501 | 5154  | 87.8272018433 | 5270  | +2%    |     (run time too short) 0.0000 |
| TPC-H 9   | 25.6956367493 | 1542  | 25.7791614532 | 1547  | +0%    |     (run time too short) 0.0113 |
| TPC-H 10  | 34.8038711548 | 2089  | 34.119392395  | 2048  | -2%    |     (run time too short) 0.0000 |
| TPC-H 11  | 214.630950928 | 10000 | 214.798858643 | 10000 | +0%    |     (run time too short) 0.5282 |
| TPC-H 12  | 64.2281951904 | 3854  | 64.9372024536 | 3897  | +1%    |     (run time too short) 0.0000 |
| TPC-H 13  | 28.1689796448 | 1691  | 27.9091014862 | 1675  | -1%    |     (run time too short) 0.0000 |
| TPC-H 14  | 189.532104492 | 10000 | 188.044448853 | 10000 | -1%    |     (run time too short) 0.0000 |
| TPC-H 15  | 85.8486938477 | 5151  | 84.327545166  | 5060  | -2%    |     (run time too short) 0.0000 |
| TPC-H 16  | 68.9160690308 | 4135  | 70.6257095337 | 4238  | +2%    |     (run time too short) 0.0000 |
| TPC-H 17  | 7.11086082458 | 427   | 4.23036527634 | 254   | -41%   |                          0.0000 |
| TPC-H 18  | 19.1919994354 | 1152  | 19.6855373383 | 1182  | +3%    |     (run time too short) 0.0000 |
| TPC-H 19  | 58.6464996338 | 3519  | 60.3418579102 | 3621  | +3%    |     (run time too short) 0.0000 |
| TPC-H 20  | 4.12152719498 | 248   | 2.79529571533 | 168   | -32%   |                          0.0000 |
| TPC-H 21  | 2.04459118843 | 123   | 1.36030769348 | 82    | -33%   |                          0.0000 |
| TPC-H 22  | 125.008743286 | 7501  | 124.485427856 | 7470  | -0%    |     (run time too short) 0.0000 |
| average   |               |       |               |       | -5%    |                                 |
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
```

I have seen that reduction in TPC-H 2 from 130 to ~105-110 yesterday on a different branch. It feels like we have a new candidate for that jittery performance behavior. My guess is that is has to do something with correlated subqueries. We don't see it for queries without correlated subqueries.

Referencing #1122 to track this.

mrks commented:
Blocked by https://github.com/hyrise/sql-parser/pull/107

mrks commented:
@lanice - Even though we now have distinct view names for TPC-H 15, this doesn't fix the problem in the multithreaded benchmark, correct?

mrzzzrm commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -267,20 +274,23 @@ void BenchmarkRunner::_warmup_query(const QueryID query_id) {
+  const auto& sql =
+      prepare ? _query_generator->preparation_queries()[query_id] : _query_generator->build_query(query_id);
+
+  if (prepare && sql.empty()) return {};
```
sql can be empty? Does this imply that every BenchmarkQueryGeneratorThing needs to provide an empty PreparationQuery for each of its queries?

mrks replied:
Yes. For example, TPC-H 15 cannot be prepared.

The alternative would be a method `bool query_needs_preparation`, but I don't see that improving the code. What we could also do is have a single, large SQL string that prepares all queries for a given benchmark.

mrzzzrm replied:
That options sounds somewhat attractive to me. Maybe `AbstractQueryGenerator::get_setup_query()` or similar?

mrzzzrm replied:
("preparation_query"; I know its derived from PREPARE; but I didn't immediately associate it with "ah, this returns PREPARE statements")

mrks replied:
Mmmh, if we go down that road, shouldn't the query generator simply prepare everything as part of its constructor? I don't see an advantage in that back-and-forth.

mrzzzrm replied:
Could do that, but I don't like constructors doing work (because there MIGHT be virtual function calls hidden in that work, you never know) and I think there is some value in the BenchmarkRunner logging something along the lines of "- Running setup query: PREPARE a, PREPARE b, ...` and clearly informing the user what is happening.

lanice replied:
To throw in my opinion here as well: I like having separate SQL strings better than one big one to be honest. And if you just split the `_execute` method and have a separate `_prepare` method as you suggested in the other thread, you can just replace the `_execute_query(query, true)` in the beginning with `_prepare_query(query)` and we're good, right?
This gets rid of the boolean parameter in the schedule and and execute methods, keeps the preparation in the BenchmarkRunner to inform the user, and is very clear to the user.

mrzzzrm commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -155,7 +162,7 @@ void BenchmarkRunner::_benchmark_permuted_query_set() {
         };
 
-        auto query_tasks = _schedule_or_execute_query(query_id, on_query_done);
+        auto query_tasks = _schedule_or_execute_query(query_id, false, on_query_done);
```
You know what I think about literal boolean parameters and that I do not know what `_schedule_or_execute_query(query_id, false, on_query_done);` does. 

Something in the spirit of `_schedule_or_execute_query(query_id, IsPreparationQuery::Yes, on_query_done);` would be better, but then I'd ask myself why a function called "_schedule_or_execute_query" should care about whether the query its supposed to schedule_or_execute is a PREPARE query or not.

mrks replied:
I was planning to write a review comment about this, but forgot.

Originally I thought that splitting `_schedule_or_execute_query` into one method for preparation and another for execution would duplicate a lot of code, but now that you made me look at it again, I realize that there is no point in scheduling the preparation queries. I'll simply add a method `_prepare_query`.

lanice replied:
We wouldn't need the `prepare` parameter in `_schedule_or_execute_query` and `_schedule_query` anyway, right? It is only called with `false` everywhere.

lanice commented:
> Even though we now have distinct view names for TPC-H 15, this doesn't fix the problem in the multithreaded benchmark, correct?

Correct. As of now, the "create view" statement query still needs to be executed before we can get the plan for the next statement.

mrks commented:
I am unsure about separate queries vs. a single one. It probably does not make that much of a difference anyway.

How do you feel about 5fc90bedf?

mrks commented:
oups

lanice commented:
I mean, like you said, it does not make so much of a difference anyway...

But I think you forgot to actually implement `_prepare_query`, no? ;)

lanice commented:
Oh well I'm kind of blind. But in my defense, you did add the declaration in the .hpp

mrks commented:
Oups, yes. Changed my mind when I saw how trivial that method would be.

lanice commented on src/benchmarklib/abstract_query_generator.cpp:
```diff
@@ -2,6 +2,8 @@
 
 namespace opossum {
 
+std::string AbstractQueryGenerator::setup_queries() const { return ""; }
```
Why did we change from `preparation_queries` to `setup_queries`?
Because this one is ambiguous. I would expect to be able to call `TPCHQueryGenerator::setup_queries()` and then my queries are set up. You could use `get_setup_queries`, but even then `get_prepare_queries` would make more sense because it returns PREPARE queries.

mrks replied:
> ("preparation_query"; I know its derived from PREPARE; but I didn't immediately associate it with "ah, this returns PREPARE statements")

`get_preparation_queries` maybe? @mrzzzrm ?

mrzzzrm replied:
if you have to.

lanice commented on src/benchmarklib/benchmark_runner.hpp:
```diff
@@ -46,14 +46,15 @@ class BenchmarkRunner {
 
   // Execute named_query
-  void _execute_query(const QueryID query_id, const std::function<void()>& done_callback);
+  void _execute_query(const QueryID query_id, const std::function<void()>& done_callback = []() {});
```
You added these defaults because of the `prepare` bool parameter you added, right? Since that is gone, I would also remove the default parameters for `done_callback` again.

mrks replied:
Mmh, I left this in on purpose, because I can't see a reason why we would want to enforce a callback.

mrks replied:
ok

mrks commented:
Oh crap. sqlite does not support the "PREPARE/EXECUTE" syntax. Basically, we have two options: Either, we do a lot of hacking so that prepared statements are run for Hyrise and transform the queries so that sqlite likes them, or we simply do our verification with the verification data provided by the TPC-H specification.

I prefer the second, because it causes less headaches. Also, that's what the verification data is for.

lanice commented:
`Also, that's what the verification data is for.`

Sounds like a good reason for me :+1:

mrzzzrm commented:
TPC-H verification data, what is that? google doesn't bring up anything on first try.

mrzzzrm commented:
If this verification data is actual data I might be fine with it. If it's just "Query X return 2*sf rows" then I am not. Having non-PREPARE/EXECUTE variants of the TPC-H queries would be preferable then.

mrks commented:
They call it validation output and it's part of the specification.

For each query, they give the first row of the expected output. Since all queries are ordered, this should be fine

mrzzzrm commented:
Expected it to be something along the lines of that. I am not sure how many rows TPCH queries typically return... But I can probably write up  hundred reasons why hyrise might produce the correct first row and incorrect/missing/additional rows after.

mrks commented:
Absolutely, but if these make it through the regular tests, we f'ed up before. In my opinion, the TPC-H tests should primarily test the correct generation of the tables and the correct feeding of the queries into the SQL pipeline.

I would also prefer it if the specification provided all results.

How about we additionally generate the current results as a tbl file and use these for validation?

mrks commented:
Ok, for the debug build, using the TPC-H validation set is utopic:

```
[ RUN      ] TPCHTestInstances/TPCHTest.TPCHQueryTest/13
[       OK ] TPCHTestInstances/TPCHTest.TPCHQueryTest/13 (7991136 ms)
```

My suggestion is to first validate that our current result match the TPC-H validation set (I spot at least one discrepancy) and then export the sqlite results once. I don't see a value in running sqlite for each test iteration if the query parameters and the scale factor remain unchanged. This way, we avoid having to rewrite the queries a second time and introducing a certain chance of errors. A useful side effect is that we save some of the execution time of hyriseSystemTest, especially when using memcheck.

mrzzzrm commented:
> Ok, for the debug build, using the TPC-H validation set is utopic:

Sorry, maybe I am stupid, but what is this "TPC-H validation set"? 

Is it this thing that you have to obtain from the TPC?

![image](https://user-images.githubusercontent.com/2679018/49004094-cc831880-f163-11e8-9925-f422a417d9c6.png)

mrzzzrm commented:
I see your point that having a second variant of the TPC-H queries in our codebase is a potential source for error... Why don't we leave `tpch_queries.cpp` as it is (and continue to use it for the TPC-H tests) and rely on qgen to generate the queries for the benchmark? Because qgen will always generate an instance for each of the 22 queries and doesn't allow us to create an instance for a selected query individually?

mrks commented:
* qgen doesn't use prepared statements.
* If qgen's performance is anywhere close to that of dbgen, we can stop worrying about ours.
* Most importantly, in my opinion, this would defeat the purpose of the TPC-H test as a whole. Primarily, we should test that we properly execute the test, i.e., that the benchmark results are close to what TPC-H expects from us. Being able to stress-test our code is a welcomed side effect. If we use different sources for the generation of the queries, we will not be test our TPC-H benchmark.

I just found the entire validation data in the tools.zip that you can download from the TPC-H site. That way, we could test the entire result for SF1 in release mode. Once we have decided what to do about expensive debug tests, we might use exported results of a smaller SF there.

mrks commented:
I looked into it and found two more issues with qgen:
* qgen writes the queries into a sql file, which we would have to parse again
* As we don't have proper date functions yet, we would need to either modify qgen to do those date calculations as part of the generation or modify the SQL strings later on.

mrzzzrm commented:
qgen can be used to generate just the parameters `-l file_for_params.txt` - If not using qgen, how would you generate meaningful parameters for the PREPAREd TPCH statements in the future?

mrzzzrm commented:
Where in the reference data do you find query results? For me, it just seems to contain subsets of the tables.

(Assuming we're talking about this:)
![image](https://user-images.githubusercontent.com/2679018/49006185-cb081f00-f168-11e8-8219-5b9fb9d41ea1.png)

mrks commented:
Ah, I was not aware of that. That would have saved some work (but would still require file and string manipulation to build our query strings, no matter if prepared or not).

My current plan is something like this: https://github.com/hyrise/hyrise/blob/5d717e85b9d1b0f38311a1bcf7a5e1ebc0ccdec3/src/benchmarklib/tpch/tpch_query_generator.cpp#L49

You'll find the reference results in dbgen/answers.

mrzzzrm commented:
@mrks are you currently not working on this, or are you waiting for input?

mrks commented:
Didn't find the time yet. Please check if what I remember from our meeting is correct:

* We will have both TPC-H compliant, non-prepared and prepared queries (default)
* We will have a `--verify` option for the non-prepared case
* The tests should use stored tbl files for verification

mrzzzrm commented:
Yes. `--verify` doesn't need to be part of this PR, though.

mrks commented:
> We will have both TPC-H compliant, non-prepared and prepared queries (default)

Implemented

>    We will have a --verify option for the non-prepared case

Moved to #741 

>    The tests should use stored tbl files for verification

Implemented. Using the TPC-H provided SF1 results is now in #741

mrzzzrm commented:
I take it this is ready to be looked at again?

mrks commented:
Yes, please

mrks commented on src/benchmarklib/tpch/tpch_query_generator.hpp:
```diff
@@ -6,12 +6,27 @@ namespace opossum {
 
 class TPCHQueryGenerator : public AbstractQueryGenerator {
```
No need for that. A follow-up PR that introduces randomness is basically ready once this is merged.

mrks commented on src/benchmarklib/tpch/tpch_query_generator.cpp:
```diff
@@ -25,9 +30,84 @@ void TPCHQueryGenerator::_generate_names() {
+    return query_15;
+  }
+
+  // Stores how the parameters (the ? in the query) should be replaced
```
I'll add a comment, but for those minimal tests, we do not use the demo values defined in the standard.

mrks replied:
// Stores how the parameters (the ? in the query) should be replaced. These values are examples for the queries. Most
  // of them use the validation parameters given in the TPC-H specification for the respective query. A few are
  // modified so that we get results even for a small scale factor.

mrks commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -50,7 +50,20 @@ BenchmarkRunner::~BenchmarkRunner() {
-  _config.out << "\n- Starting Benchmark..." << std::endl;
+  // Run the preparation queries
+  {
+    const auto& sql = _query_generator->get_preparation_queries();
```
This is not nice, and I will change it to simply `auto`, but it is not wrong. It's sometimes referred to as Temporary (Object|Reference) Lifetime Extension. Here is an article that might be of interest: https://abseil.io/tips/107

mrks commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -50,7 +50,20 @@ BenchmarkRunner::~BenchmarkRunner() {
+  {
+    const auto& sql = _query_generator->get_preparation_queries();
+
+    if (!sql.empty()) {
```
Mmmh, I feel like that would overcomplicate things. How about a simple comment "Some benchmarks might not need preparation"?

mrzzzrm replied:
should do it

mrks commented on src/benchmarklib/tpch/tpch_query_generator.cpp:
```diff
@@ -25,9 +30,84 @@ void TPCHQueryGenerator::_generate_names() {
+    // Generating TPC-H Query 15 by hand
+    auto query_15 = std::string{tpch_queries.find(15)->second};
+
+    static auto view_id = 0;
```
// TPC-H query 15 uses "stream ids" to name the views. While not supported right now, we might want to execute
    // multiple instances of Q15 simultaneously and will need unique view names for that.

mrks commented on src/benchmarklib/tpch/tpch_query_generator.cpp:
```diff
@@ -25,9 +30,84 @@ void TPCHQueryGenerator::_generate_names() {
+  return _build_query_with_placeholders(query_id, parameter_values[query_id]);
+}
+
+std::string TPCHQueryGenerator::_build_query_with_placeholders(const QueryID query_id,
```
`_build_parameterized_query`?

mrzzzrm replied:
Still sounds more like `SELECT * FROM t WHERE a > ?` than like `SELECT * FROM t WHERE a > 5`, does it not? I think `_build_executable_query()` contrasts this function better to the function that creates "preparation" queries.

mrks commented on src/test/tables/tpch/test-validation/q9.tbl:
```diff
@@ -0,0 +1,27 @@
+nation|o_year|sum_profit
```
"These files are previous results of a known-to-be-good (i.e., validated with SQLite) execution of the tests." ?

mrks replied:
I'd rather add that comment to the line where the files are accessed.

mrzzzrm replied:
both fine with me.

mrks commented:
Four remaining things where I'd like some input

mrzzzrm commented:
Did I get them all?
